pkg/status/status.go

```diff
@@ -199,3 +199,12 @@ func ShouldRetryUpgrade(hr *v1.HelmRelease) bool {
 	}
 	return hr.Spec.Rollback.GetMaxRetries() == 0 || hr.Status.RollbackCount <= hr.Spec.Rollback.GetMaxRetries()
 }
+
+// ShouldRetryInstall returns if the install of a failed release should
+// be retried.
+func ShouldRetryInstall(hr *v1.HelmRelease) bool {
+	if !HasSynced(hr) {
+		return true
+	}
+	return hr.GetMaxRetries() == 0 || hr.Status.InstallCount < hr.GetMaxRetries()
```

Comment:
```suggestion
	return hr.GetMaxRetries() < 0 || hr.Status.InstallCount < hr.GetMaxRetries()
```

It would be useful to be able to do 0 retries. To accomplish this, a solution could be to deprecate `rollback.retry` and change default to true, change `rollback.maxRetries: 0` to mean 0 retries, and `rollback.maxRetries: -1` to mean infinite. Same behavior goes for the root-level `maxRetries`, which serves as the default value for `rollback.maxRetries`.

Reply:
I was avoiding any changes that involved changing anything around rollback, but I see no issue to use 0 and -1 for the root level max retries to turn them off or be infinite.

Reply:
After a little more thought, an `uninstall` field could be more consistent with the existing `rollback` field, with the exception of a couple defaults being different:

```yaml
spec:
  uninstall:
    enable: true # default true
    retry: true # default true
    maxRetries: 1 # default 5
  rollback: # no changes
    enable: true # default false
    retry: true # default false
    maxRetries: 1 # default 5
```

and then instead of `status.installCount` it could be `status.uninstallCount` to match `status.rollbackCount`.

Reply:
I like the above proposal, and the spec fits in to the picture I had in mind for a future gen API with finer grain option control.

Reply:
Do the semantics of "uninstall" make sense to a user however? I still think of it as retrying the install and not the uninstalls, in my mind I care about retrying the install if it fails, not the uninstall.

Reply:
> And when say `uninstall.enable=false` do we now not uninstall when removing a HR for example?

We do now uninstall when an HR is removed, and I would expect this behaviour to stay put. Due to this, I agree "install" is a better option as the flag and logic applies to this process.

Reply:
We said we cannot use "install" right? I know when we talked about it you had said we cannot as it is seen as the spec root. I had not tried it yet but just recall that.

Reply:
Okies after chatting with @hiddeco offline the plan is to add the retry related fields into an `install` section.

This change will also pave the way to move other "install" related fields into a logical grouping in the future moving them out of the spec root.

Thoughts on that plan @seaneagan ?

```
spec:
  install:
    retry: true # default true
    maxRetries: 1 # default 5
  rollback: # no changes
    enable: true # default false
    retry: true # default false
```

Reply:
Yes, I agree that it is really about the desire to retry installs, not to do uninstalls, and install retries should imply uninstalls. But similarly some of the rollback fields are about the desire to retry upgrades, so I think similarly it would make sense to deprecate those fields and replace them with an `upgrade` key, which I think could be done in a backward compatible way. And if moving them, should consider removing the `retry` field, and just allowing `maxRetries=0` to mean no retries. Also it may be nice to be able to define whether to do a rollback/uninstall after all retries are consumed (or leave release in failed state for debugging).

```yaml
spec:
  install:
    maxRetries: 1 # default 0? (none), >0 implies uninstalls
    uninstallAfterMaxRetries: true # default false, whether to uninstall after max retries consumed
consumed
    noHooks: true
  upgrade:
    maxRetries: 1 # default 0 (none), >0 implies rollbacks
    rollbackAfterMaxRetries: true # default false, whether to rollback after max retries consumed
    force: true
    noHooks: true
  rollback: # how to do rollbacks, not whether to
    force: true
    noHooks: true
    timeout: 300
  uninstall: # how to do uninstalls, not whether to 
    noHooks: true
    timeout: 300
```

`rollback.enable/retry/maxRetries` would all be deprecated, and it would be an error to specify any of them at the same time as specifying an `upgrade` key. It may make sense to also deprecate `status.rollbackCount` for `status.upgradeRetryCount` and add `status.installRetryCount`.

Reply:
@seaneagan that all seems sensible, but definately something I won't tackle as part of this PR, it has already grown a bit hairier than I anticipated in some of the edge cases but I can definately see a future PR after this to go through everything and improve the semantics behind all the existing fields.